stats/stats.go

```diff
@@ -0,0 +1,253 @@
+
+		// if we have reach end of file, just return.
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return nil, err
+		}
+
+		logLine := string(line)
+		if !messageHeader.MatchString(logLine) {
+			continue
+		}
+
+		dataLog, err := decodeLogLine(logLine)
+		if err != nil {
+			continue
+		}
+
+		var message common.MessageState
+		if err := json.Unmarshal([]byte(dataLog), &message); err != nil {
+			continue
+		}
+
+		msgAggr, ok := aggregates[message.Hash]
+		if !ok {
+			msgAggr.Envelope = message.Hash
+			msgAggr.Direction = message.Type
+			msgAggr.Protocol = message.Protocol
+			msgAggr.Sent = time.Unix(int64(message.TimeSent), 0)
+			msgAggr.Payload = string(message.Payload)
+			msgAggr.Request = message.Source
+			msgAggr.FromDevice = message.FromDevice
+			msgAggr.ToDevice = message.ToDevice
+
+			order = append(order, message.Hash)
+		}
+
+		// This should be impossible, but skip any whoes Envelope hash matches but
+		// has different protocols.
+		if ok && msgAggr.Protocol != message.Protocol && msgAggr.Envelope != message.Hash {
+			continue
+		}
+
+		if message.Type == outgoingMessage {
+			if msgAggr.ToDevice == "" {
+				msgAggr.ToDevice = message.ToDevice
+			}
+
+			if msgAggr.FromDevice == "" {
+				msgAggr.FromDevice = message.FromDevice
+			}
+
+			if msgAggr.Payload == "" {
+				msgAggr.Payload = string(message.Payload)
+			}
+		}
+
+		switch message.Status {
+		case "Pending":
+			if message.Type == outgoingMessage {
+				msgAggr.Request = message.Source
+			}
+
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Pending",
+				When:   message.Received,
+			})
+		case "Sent":
+			if message.Type == outgoingMessage {
+				msgAggr.Request = message.Source
+			}
+
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Sent",
+				When:   message.Received,
+			})
+		case "Resent":
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Resent",
+				When:   message.Received,
+			})
+		case "Queued":
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Queued",
+				When:   message.Received,
+			})
+		case "Cached":
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Cached",
+				When:   message.Received,
+			})
+		case "Delivered":
+			if msgAggr.ToDevice == "" {
+				msgAggr.ToDevice = message.ToDevice
+			}
+
+			if msgAggr.FromDevice == "" {
+				msgAggr.FromDevice = message.FromDevice
+			}
+
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Delivered",
+				When:   message.Received,
+			})
+		case "Rejected":
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Rejected",
+				When:   message.Received,
+				Error:  message.RejectionError,
+			})
+		case "Processing":
+			msgAggr.Timeline = append(msgAggr.Timeline, Timeline{
+				Status: "Processing",
+				When:   message.Received,
+			})
+		}
+
+		aggregates[message.Hash] = msgAggr
+	}
+
+	var messages []AggregatedMessage
```

Comment:
If such cases, you can set slice's size during initialization: `messages := make([]AggregatedMessage, 0, len(order))`.